File: day-11/prg.php
<?php /** @noinspection PhpUnhandledExceptionInspection */

class Instruction
{
    private $code;

    public function __construct(int $code)
    {
        $this->code = $code;
    }

    public function opCode(): int
    {
        return $this->code % 100;
    }

    public function mode($i): int
    {
        $modes = intdiv($this->code, 100);
        return intdiv($modes, 10 ** ($i - 1)) % 10;
    }
}

class State
{
    private $ip;
    private $rb;
    private $mem;
    private $input;
    private $output;

    public function __construct(int $ip, int $rb, array $mem, callable $input, array $output)
    {
        $this->ip = $ip;
        $this->rb = $rb;
        $this->mem = $mem;
        $this->input = $input;
        $this->output = $output;
    }

    public function decode(): Instruction
    {
        return new Instruction($this->mem[$this->ip]);
    }

    private function access(int $i): int
    {
        return array_key_exists($i, $this->mem) ? $this->mem[$i] : 0;
    }

    public function argument(int $i, int $mode): int
    {
        $val = $this->access($this->ip + $i);
        switch ($mode) {
            case 0:
                return $this->access($val);
            case 1:
                return $val;
            case 2:
                return $this->access($val + $this->rb);
            default:
                throw new Exception('Unexpected value ' . $mode);
        }
    }

    public function read(): int
    {
        return ($this->input)();
    }

    public function copy(): State
    {
        return new State($this->ip, $this->rb, $this->mem, $this->input, $this->output);
    }

    public function store(int $i, int $mode, int $value): State
    {
        $val = $this->access($this->ip + $i);
        switch ($mode) {
            case 0:
                $this->mem[$val] = $value;
                break;
            case 2:
                $this->mem[$val + $this->rb] = $value;
                break;
            default:
                throw new Exception('Unexpected value ' . $mode);
        }
        return $this;
    }

    public function adjustRelBase(int $adj): State
    {
        $this->rb += $adj;
        return $this;
    }

    public function setIp(int $ip): State
    {
        $this->ip = $ip;
        return $this;
    }

    public function moveIp(int $move): State
    {
        $this->ip += $move;
        return $this;
    }

    public function write($value): State
    {
        $this->output[] = $value;
        return $this;
    }

    public function halted(): bool
    {
        return $this->ip < 0;
    }

    public function peekOutputs(): array
    {
        return $this->output;
    }

    public function consumeOutputs(): array
    {
        $out = $this->output;
        $this->output = [];
        return $out;
    }
}

interface Instr
{
    public function perform(Instruction $instruction, State $state): State;
}

class Add implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        return $state->copy()->store(3, $instruction->mode(3), $a1 + $a2)->moveIp(4);
    }
}

class Mul implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        return $state->copy()->store(3, $instruction->mode(3), $a1 * $a2)->moveIp(4);
    }
}

class Inp implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->read();
        return $state->copy()->store(1, $instruction->mode(1), $a1)->moveIp(2);
    }
}

class Outp implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        return $state->copy()->write($a1)->moveIp(2);
    }
}

class JmpT implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        if ($a1 !== 0) {
            return $state->copy()->setIp($a2);
        } else {
            return $state->copy()->moveIp(3);
        }
    }
}

class JmpF implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        if ($a1 === 0) {
            return $state->copy()->setIp($a2);
        } else {
            return $state->copy()->moveIp(3);
        }
    }
}

class Lt implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        return $state->copy()->store(3, $instruction->mode(3), $a1 < $a2 ? 1 : 0)->moveIp(4);
    }
}

class Eq implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        $a2 = $state->argument(2, $instruction->mode(2));
        return $state->copy()->store(3, $instruction->mode(3), $a1 === $a2 ? 1 : 0)->moveIp(4);
    }
}

class ARb implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        $a1 = $state->argument(1, $instruction->mode(1));
        return $state->copy()->adjustRelBase($a1)->moveIp(2);
    }
}

class Halt implements Instr
{
    public function perform(Instruction $instruction, State $state): State
    {
        return $state->copy()->setIp(-1);
    }
}

class Computer
{
    private $state;
    private $instrs;

    public function __construct(array $mem, callable $input)
    {
        $this->state = new State(0, 0, $mem, $input, []);
        $this->instrs = [
            1 => new Add(),
            2 => new Mul(),
            3 => new Inp(),
            4 => new Outp(),
            5 => new JmpT(),
            6 => new JmpF(),
            7 => new Lt(),
            8 => new Eq(),
            9 => new ARb(),
            99 => new Halt()
        ];
    }

    function step(): State
    {
        $instruction = $this->state->decode();
        $opCode = $instruction->opCode();

        if (!array_key_exists($opCode, $this->instrs)) {
            throw new Exception("Invalid op code: " . $opCode);
        }

        $instr = $this->instrs[$opCode];
        $newState = $instr->perform($instruction, $this->state);

        $this->state = $newState;
        return $newState;
    }

    function state(): State
    {
        return $this->state;
    }

    function loopWhile(callable $condition): State
    {
        while (!$this->state->halted() && $condition($this->state)) {
            $this->step();
        }
        return $this->state;
    }

    function halted(): bool
    {
        return $this->state->halted();
    }
}

function readMemory(string $file)
{
    return array_map(
        function ($token) {
            return intval($token);
        },
        explode(",", file_get_contents($file))
    );
}

class Position
{
    private $x;
    private $y;
    private $dx;
    private $dy;

    public function __construct($x, $y, $dx = 0, $dy = -1)
    {
        $this->x = $x;
        $this->y = $y;
        $this->dx = $dx;
        $this->dy = $dy;
    }

    function leftAndGo(): Position
    {
        return new Position($this->x + $this->dy, $this->y - $this->dx, $this->dy, -$this->dx);
    }

    function rightAndGo(): Position
    {
        return new Position($this->x - $this->dy, $this->y + $this->dx, -$this->dy, $this->dx);
    }

    public function __toString()
    {
        return "$this->x,$this->y";
    }
}

class Painter
{
    private $comp;
    private $hull;
    private $position;

    public function __construct(array $mem, array $hull = [])
    {
        $this->position = new Position(0, 0);
        $this->comp = new Computer($mem, $this);
        $this->hull = $hull;
    }

    public function __invoke() // computer reads
    {
        $key = strval($this->position);
        $color = array_key_exists($key, $this->hull) ? $this->hull[$key] : 0;
#        echo "Read $color at $this->position\n";
        return $color;
    }

    public function step()
    {
        $state = $this->comp->loopWhile(function (State $st) {
            return count($st->peekOutputs()) < 2;
        });
        $out = $state->consumeOutputs();

        if (array_key_exists(0, $out)) {
            $key = strval($this->position);
            $color = $out[0];
            $this->hull[$key] = $color;
#            echo "Painted hull $color at $this->position\n";
        }

        if (array_key_exists(1, $out)) {
            $dir = $out[1];
#            echo "Moving $dir at $this->position\n";
            switch ($dir) {
                case 0:
                    $this->position = $this->position->leftAndGo();
                    break;
                case 1:
                    $this->position = $this->position->rightAndGo();
                    break;
                default:
                    throw new Exception('Invalid direction ' . $dir);
            }
        }
    }

    public function loop()
    {
        while (!$this->comp->halted()) {
            $this->step();
        }
    }

    public function hull(): array
    {
        return $this->hull;
    }
}

function printHull(array $hull)
{
    $minx = PHP_INT_MAX;
    $miny = PHP_INT_MAX;
    $maxx = PHP_INT_MIN;
    $maxy = PHP_INT_MIN;
    foreach ($hull as $k => $v) {
        list($x, $y) = explode(",", $k);
        $x = intval($x);
        $y = intval($y);
        $minx = min($minx, $x);
        $miny = min($miny, $y);
        $maxx = max($maxx, $x);
        $maxy = max($maxy, $y);
    }
    $paint = [];
    for ($y = $miny; $y <= $maxy; $y++) {
        $paint[$y - $miny] = [];
        for ($x = $minx; $x <= $maxx; $x++) {
            $paint[$y - $miny][$x - $minx] = '.';
        }
    }
    foreach ($hull as $k => $v) {
        list($x, $y) = explode(",", $k);
        $x = intval($x);
        $y = intval($y);
        $paint[$y - $miny][$x - $minx] = $v != 0 ? '#' : '.';
    }

    echo implode("\n", array_map(function ($row) {
        return implode("", $row);
    }, $paint));
}

function paint($hull)
{
    $painter = new Painter(readMemory("input"), $hull);
    $painter->loop();
    return $painter->hull();
}

function tasks() {
    $hull1 = paint([]);
    echo "Estimate: " . count($hull1) . "\n";

    $hull2 = paint(["0,0"=>1]);
    printHull($hull2);
}

tasks();

File: day-13/prg.rb
def read_memory(file_name)
  str = IO.read(file_name).strip
  str.split(",").map { |t| t.to_i }
end

class Operation
  attr_accessor :op

  def initialize(op)
    @op = op % 100
    @modes = op / 100
  end

  def mode(i)
    @modes / (10 ** (i - 1)) % 10
  end
end

class State
  def initialize(ip, rb, mem, input, output)
    @ip = ip
    @rb = rb
    @mem = mem
    @input = input
    @output = output
  end

  def decode
    Operation.new(@mem[@ip])
  end

  def argument(op, i)
    a = @mem[@ip + i]
    mode = op.mode(i)
    case mode
    when 0
      @mem[a]
    when 1
      a
    when 2
      @mem[a + @rb]
    else
      raise "Unknown argument mode #{mode}"
    end
  end

  def input
    @input.call(self)
  end

  def dup
    #State.new(@ip, @rb, @mem.dup, @input, @output)
    self
  end

  def store(op, i, val)
    a = @mem[@ip + i]
    mode = op.mode(i)
    case mode
    when 0
      @mem[a] = val
    when 2
      @mem[a + @rb] = val
    else
      raise "Unknown store mode #{mode}"
    end
    self
  end

  def move_by(dip)
    @ip += dip
    self
  end

  def move_to(ip)
    @ip = ip
    self
  end

  def adjust_rb(drb)
    @rb += drb
    self
  end

  def output(val)
    @output.call(val)
    self
  end

  def halted?
    @ip < 0
  end

  def to_s
    "ip=#{@ip}, rb=#{@rb}, output=#{@output}, mem=#{@mem}"
  end

  def hack(i)
    @mem[i]
  end
end

class Processor
  def add(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    state.dup.store(op, 3, a1 + a2).move_by(4)
  end

  def mul(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    state.dup.store(op, 3, a1 * a2).move_by(4)
  end

  def inp(state, op)
    a1 = state.input
    if a1 == nil
      raise "No input"
    end
    state.dup.store(op, 1, a1).move_by(2)
  end

  def outp(state, op)
    a1 = state.argument(op, 1)
    state.dup.output(a1).move_by(2)
  end

  def jmpt(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    if a1 != 0
      state.dup.move_to(a2)
    else
      state.dup.move_by(3)
    end
  end

  def jmpf(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    if a1 == 0
      state.dup.move_to(a2)
    else
      state.dup.move_by(3)
    end
  end

  def lt(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    state.dup.store(op, 3, a1 < a2 ? 1 : 0).move_by(4)
  end

  def eq(state, op)
    a1 = state.argument(op, 1)
    a2 = state.argument(op, 2)
    state.dup.store(op, 3, a1 == a2 ? 1 : 0).move_by(4)
  end

  def arb(state, op)
    a1 = state.argument(op, 1)
    state.dup.adjust_rb(a1).move_by(2)
  end

  def halt(state, op)
    state.dup.move_to(-1)
  end

  def step(state, op)
    case op.op
    when 1
      add(state, op)
    when 2
      mul(state, op)
    when 3
      inp(state, op)
    when 4
      outp(state, op)
    when 5
      jmpt(state, op)
    when 6
      jmpf(state, op)
    when 7
      lt(state, op)
    when 8
      eq(state, op)
    when 9
      arb(state, op)
    when 99
      halt(state, op)
    else
      raise "Unknown operation #{op.op}"
    end
  end
end

class Computer
  attr_accessor :state, :output

  def initialize(mem, input)
    m = Hash.new(0)
    mem.each_with_index do |val, index|
      m[index] = val
    end

    @output = []

    @state = State.new(0, 0, m, input, lambda { |val| @output.push(val) })
    @processor = Processor.new
  end

  def step
    op = @state.decode
    @state = @processor.step(@state, op)
    self
  end

  def steps
    while !halted? && (!block_given? || (yield @state))
      step
    end
    self
  end

  def halted?
    @state.halted?
  end

  def to_s
    "halted=#{halted?}, state=#{@state}"
  end
end

TILES = [' ', '#', '@', '_', 'o']

def tile(output)
  minx = 1000
  maxx = -1000
  miny = 1000
  maxy = -1000
  output.each_slice(3) do |x, y|
    if x != -1
      minx = minx < x ? minx : x
      maxx = maxx > x ? maxx : x
      miny = miny < y ? miny : y
      maxy = maxy > y ? maxy : y
    end
  end

  a = Array.new(maxy - miny + 1) { Array.new(maxx - minx + 1) { ' ' } }

  output.each_slice(3) do |x, y, t|
    a[y - miny][x - minx] = TILES[t]
  end
  a.map { |aa| aa.join('') }.join("\n")
end

def score(output)
  s = -1
  output.each_slice(3) do |x, y, t|
    if x == -1 && t > s
      s = t
    end
  end
  s
end

def task1(mem)
  comp = Computer.new(mem, [])
  comp.steps { |x| true }

  t = tile comp.output
  puts t
  puts t.count("@")
end

def task2(mem)
  mem2 = mem.dup
  mem2[0] = 2

  pad = 20
  comp = Computer.new(mem2, lambda { |st|
    bx = st.hack(388)

    dx = 0
    if bx < pad
      pad -= 1
      dx = -1
    elsif bx == pad
      0
    else
      pad += 1
      dx = 1
    end
    dx
  })
  comp.steps { |x| true }

  puts tile(comp.output)
  puts score(comp.output)
end

mem = read_memory("input")

task1(mem)
task2(mem)
